cpp/src/gandiva/function_registry.cc

```diff
@@ -16,18 +16,23 @@
 // under the License.
 
 #include "gandiva/function_registry.h"
+
+#include <iterator>
+#include <utility>
+#include <vector>
+
+#include "arrow/util/logging.h"
 #include "gandiva/function_registry_arithmetic.h"
 #include "gandiva/function_registry_datetime.h"
 #include "gandiva/function_registry_hash.h"
 #include "gandiva/function_registry_math_ops.h"
 #include "gandiva/function_registry_string.h"
 #include "gandiva/function_registry_timestamp_arithmetic.h"
 
-#include <iterator>
-#include <utility>
-#include <vector>
-
 namespace gandiva {
+constexpr uint32_t kMaxFunctionSignatures = 2048;
+
+FunctionRegistry::FunctionRegistry() { pc_registry_.reserve(kMaxFunctionSignatures); }
 
 FunctionRegistry::iterator FunctionRegistry::begin() const {
```

Comment:
There is one caveat here, the `pc_registry_map_` stores the signatures with their pointers and key/values:
```
  for (auto& elem : pc_registry_) {
    for (auto& func_signature : elem.signatures()) {
      map.emplace(&func_signature, &elem);
    }
}
```

It is not a problem previously if the size of function registry is known up front, however, here we allows users to add new functions, which makes the `pc_registry_` std::vector to be resized, and the pointer stored in `pc_registry_map_` will be invalidated after `pc_registry_` is resized.

Initially, I would like to switch `pc_registry_` from `std::vector` to `std::list` to guarantee pointer stability, however, I find currently `FunctionRegistry` provides an `iterator` API `begin/end/back` which returns `iterator = const NativeFunction *`, this assumes the backend container is `std::vector`, which stores these `NativeFunction` in a continuous manner.  This iterator API is used by `ExpressionRegistry`, which is likely used by applications (e.g. the `GetRegisteredFunctionSignatures` is never used internally in Gandiva project). 

To avoid breaking existing API, currently, I have to pre-allocate the `pc_registry_` vector so that it won't be resized, but this limits the registry to have capacity 2048 (which is likely enough in most cases). Please advise if there is better approach for addressing this issue. Thanks.

Reply:
Well, we can keep the `GetRegisteredFunctionSignatures` API as-is but still redefine `begin()` and `end()` if we want.

The question is whether some people rely on the fact that `begin()` and `end()` return a pointer, rather than a more general bidirectional iterator. That would be slight weird IMHO.

That said, I think it's also ok for now to pre-allocate the vector to workaround this problem.

Reply:
`GetRegisteredFunctionSignatures()` is used by pyarrow: https://github.com/apache/arrow/blob/4bbd48db8f3937c3e2f54f4c1131fd6af5a6b85c/python/pyarrow/gandiva.pyx#L693-L709

How about making the existing API deprecated, keeping backward compatibility and creating a new API for now? (We can do this in a separated PR.)
We can remove the deprecated existing API later.

Reply:
> How about making the existing API deprecated, keeping backward compatibility and creating a new API for now?

That makes sense, I will give it a try. Thanks.

Reply:
> How about making the existing API deprecated, keeping backward compatibility and creating a new API for now? > (We can do this in a separated PR.)
> We can remove the deprecated existing API later.

I took a closer look, and the `GetRegisteredFunctionSignatures` API could be kept easily without breaking compatibility, however, it is not easy to keep the iterator API of `FunctionRegistry` compatible while not making its internals too complicated. I may give it a try in a separated PR if I figure out a clean way to do it.

Reply:
The iterator API doesn't look very useful either, so I'm not sure it's worth trying to emulate it. Why would someone iterate over registered functions one by one?

We could simply expose an API returning a `std::vector` of the registered functions...